patina_dxe_core/src/config_tables/memory_attributes_table.rs

```diff
@@ -157,83 +225,47 @@ pub fn core_install_memory_attributes_table() {
     }
 
     // this allocates memory to do the collect, but that's okay because it is boot services memory
-    let mat_desc_list: Vec<efi::MemoryDescriptor> = desc_list
+    let mat_desc_list: Vec<MemoryDescriptor> = desc_list
         .iter()
         .filter_map(|descriptor| {
             // we only want the EfiRuntimeServicesCode and EfiRuntimeServicesData sections in the MAT
             match descriptor.r#type {
-                efi::RUNTIME_SERVICES_CODE | efi::RUNTIME_SERVICES_DATA => {
-                    Some(efi::MemoryDescriptor {
-                        attribute: MemoryProtectionPolicy::apply_memory_attributes_table_policy(
-                            descriptor.attribute,
-                            descriptor.r#type,
-                        ),
-                        // use all other fields from the GCD descriptor
-                        ..*descriptor
-                    })
-                }
+                efi::RUNTIME_SERVICES_CODE | efi::RUNTIME_SERVICES_DATA => Some(MemoryDescriptor::new(
+                    descriptor.r#type,
+                    descriptor.physical_start,
+                    descriptor.virtual_start,
+                    descriptor.number_of_pages,
+                    MemoryProtectionPolicy::apply_memory_attributes_table_policy(
+                        descriptor.attribute,
+                        descriptor.r#type,
+                    ),
+                )),
                 _ => None,
             }
         })
         .collect();
 
-    // allocate memory for the MAT and publish it
-    let buffer_size =
-        mat_desc_list.len() * size_of::<efi::MemoryDescriptor>() + size_of::<efi::MemoryAttributesTable>();
-    match core_allocate_pool(efi::BOOT_SERVICES_DATA, buffer_size) {
-        Err(err) => {
-            log::error!("Failed to allocate memory for the MAT! Status {err:#X?}");
+    let mat = MemoryAttributesTable::from_mat_desc_list(mat_desc_list);
+    let mat_ptr = mat.as_ptr();
+
+    match core_install_configuration_table(efi::MEMORY_ATTRIBUTES_TABLE_GUID, mat_ptr as *mut c_void, st) {
```

Comment:
Here you are putting `mat_ptr` into a public table (i.e. one that is visible outside of the patina core to e.g. drivers). However, because you don't leak the mat, I think it will get auto-dropped at the end of the function, meaning that the pointer you installed in the global config table is going to point to freed memory.